library/elliptics.h

```diff
@@ -58,30 +58,59 @@ struct dnet_group;
 struct dnet_net_state;
 struct dnet_cmd_stats;
 struct dnet_access_context;
+struct n2_net_iovec;
+struct n2_request_info;
+struct n2_response_info;
+
+// Define which fields of dnet_io_req are used
+enum dnet_io_req_type {
+	DNET_IO_REQ_OLD_PROTOCOL = 0, // deprecated old-protocol-dependent fields
+	DNET_IO_REQ_TYPED_REQUEST = 1, // dnet_io_req::request_info
+	DNET_IO_REQ_TYPED_RESPONSE = 2, // dnet_io_req::response_info
+};
 
 struct dnet_io_req {
-	struct list_head	req_entry;
+	struct list_head		req_entry;
 
-	struct dnet_net_state	*st;
+	struct dnet_net_state		*st;
 
-	void			*header;
-	size_t			hsize;
+	// TODO: remove deprecated fields after refactoring complete
+	void				*header; /*Deprecated*/
+	size_t				hsize; /*Deprecated*/
 
-	void			*data;
-	size_t			dsize;
+	void				*data; /*Deprecated*/
+	size_t				dsize; /*Deprecated*/
 
-	int			on_exit;
-	int			fd;
-	off_t			local_offset;
-	size_t			fsize;
+	int				on_exit; /*Deprecated*/
+	int				fd; /*Deprecated*/
+	off_t				local_offset; /*Deprecated*/
+	size_t				fsize; /*Deprecated*/
+
+	struct timespec			queue_start_ts;
+	uint64_t			queue_time;
+	uint64_t			recv_time;
+
+	struct dnet_access_context 	*context;
 
-	struct timespec		queue_start_ts;
-	uint64_t		queue_time;
-	uint64_t		recv_time;
+	// Protocol-independent members:
+	enum dnet_io_req_type		io_req_type;
+	struct n2_request_info		*request_info;
+	struct n2_response_info		*response_info;
+
+	// Temporary old-protocol send-list members: struct dnet_io_req is currently used both in request_queue
+	// and send_list. Send_list is protocol-specific and contains serialized messages which are ready to
+	// send by net. On transitional period we keep old send_list, but store serialized data in new substructure.
+	// TODO: move it to protocol-specific send_list when protocol logic'll be fully extracted
+	struct n2_net_iovec		*iov;
 
-	struct dnet_access_context *context;
 };
 
+// Because of variability of dnet_io_req, dnet_cmd can be hold there differently. Here the accessors:
+// * accessor to cmd copy (to read info)
+struct dnet_cmd *dnet_io_req_get_cmd(struct dnet_io_req *r);
+// * direct access to cmd in the request (to patch request's cmd fields)
+struct dnet_cmd *dnet_io_req_get_request_cmd_inplace(struct dnet_io_req *r);
```

Comment:
There shouldn't be direct access to cmd

Reply:
Fixed, I introduced function for changing backend_id.